File: notifications/src/main/java/io/castled/notifications/trigger/EventFilterDeserializer.java
package io.castled.notifications.trigger;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import io.castled.notifications.trigger.enums.FilterType;
import io.castled.notifications.trigger.enums.JoinType;
import io.castled.notifications.trigger.enums.OperationType;
import io.castled.notifications.trigger.enums.PropertyType;
import io.castled.notifications.trigger.models.BetweenOperation;
import io.castled.notifications.trigger.models.EventFilter;
import io.castled.notifications.trigger.models.NestedEventFilter;
import io.castled.notifications.trigger.models.PropertyFilter;
import io.castled.notifications.trigger.models.PropertyOperation;
import io.castled.notifications.trigger.models.SingleValueOperation;

public class EventFilterDeserializer implements JsonDeserializer<EventFilter> {

    @Override
    public EventFilter deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        return deserializeEventFilter(json);
    }

    private EventFilter deserializeEventFilter(JsonElement json) {
        JsonObject jsonObject = json.getAsJsonObject();
        FilterType filterType = FilterType.valueOf(jsonObject.get("type").getAsString());
        if (filterType == FilterType.NESTED) {
            return deserializeNestedFilter(json);
        } else if (filterType == FilterType.HAVING_PROPERTY) {
            return deserializePropertyFilter(json);
        } else {
            throw new JsonParseException("Unrecognized filter type!");
        }
    }

    private NestedEventFilter deserializeNestedFilter(JsonElement json) {
        JsonObject jsonObject = json.getAsJsonObject();
        JoinType joinType = JoinType.valueOf(jsonObject.get("joinType").getAsString());
        JsonArray nestedFiltersArray = jsonObject.get("nestedFilters").getAsJsonArray();

        List<EventFilter> eventFilters = new ArrayList<>();
        for (JsonElement element : nestedFiltersArray) {
            eventFilters.add(deserializeEventFilter(element));
        }
        return new NestedEventFilter(joinType, eventFilters);
    }

    private PropertyFilter deserializePropertyFilter(JsonElement json) {
        JsonObject jsonObject = json.getAsJsonObject();
        String name = jsonObject.get("name").getAsString();
        JsonElement operationElement = jsonObject.get("operation");
        PropertyOperation propertyOperation = deserializeOperation(operationElement);
        return new PropertyFilter(name, propertyOperation);
    }

    private PropertyOperation deserializeOperation(JsonElement json) {
        JsonObject jsonObject = json.getAsJsonObject();
        OperationType operationType = OperationType.valueOf(jsonObject.get("type").getAsString());
        PropertyType propertyType = PropertyType.valueOf(jsonObject.get("propertyType").getAsString());
        switch (operationType) {
            case EQ:
            case GT:
            case LT:
            case GTE:
            case LTE:
            case NEQ:
                String value = jsonObject.get("value").getAsString();
                return new SingleValueOperation(value, operationType, propertyType);
            case BETWEEN:
                String from = jsonObject.get("from").getAsString();
                String to = jsonObject.get("to").getAsString();
                return new BetweenOperation(from, to, operationType, propertyType);
            default:
                throw new JsonParseException("Unrecognized operation type!");
        }
    }
}

File: notifications/src/main/java/io/castled/notifications/trigger/EventFilterDeserializer.kt
package io.castled.notifications.trigger

import com.google.gson.JsonParseException
import io.castled.notifications.trigger.enums.FilterType
import io.castled.notifications.trigger.enums.JoinType
import io.castled.notifications.trigger.enums.OperationType
import io.castled.notifications.trigger.enums.PropertyType
import io.castled.notifications.trigger.models.*
import io.castled.notifications.trigger.models.BetweenOperation
import io.castled.notifications.trigger.models.PropertyOperation
import io.castled.notifications.trigger.models.SingleValueOperation
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.CompositeDecoder
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.encoding.decodeStructure
import kotlinx.serialization.json.*

@Serializer(forClass = EventFilter::class)
internal object EventFilterDeserializer : KSerializer<EventFilter> {


    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("EventFilter") {
        element<String>("type")
        element("GroupFilter", buildClassSerialDescriptor("GroupFilter") {
            element<JoinType>("joinType")
            element<List<EventFilter>>("filters")
        })
        element("PropertyFilter", buildClassSerialDescriptor("PropertyFilter") {
            element<String>("name")
            element<PropertyOperation>("operation")
        })
    }

    override fun deserialize(decoder: Decoder): EventFilter {
        return decoder.decodeStructure(descriptor) {
            lateinit var eventFilter: EventFilter
            while (true) {
                when (val index = decodeElementIndex(descriptor)) {
                    CompositeDecoder.DECODE_DONE -> break
                    0 -> {
                        when (decodeStringElement(descriptor, index)) {
                            FilterType.GROUP.name -> eventFilter = decodeSerializableElement(descriptor, 1, GroupFilter.serializer())
                            FilterType.HAVING_PROPERTY.name -> eventFilter = decodeSerializableElement(descriptor, 2, PropertyFilter.serializer())
                        }
                    }
                    else -> error("Unexpected index: $index")
                }
            }
            eventFilter
        }
    }

    /*
    override fun deserialize(decoder: Decoder): EventFilter {
        val jsonDecoder = decoder as? JsonDecoder
            ?: throw SerializationException("This serializer can only be used with JSON format")
        val json = jsonDecoder.decodeJsonElement()
        return deserializeEventFilter(json)
    }
    */


    override fun serialize(encoder: Encoder, value: EventFilter) {
        TODO("Not yet implemented")
    }

    private fun deserializeEventFilter(json: JsonElement): EventFilter {
        val filterType = (json.jsonObject["type"] as JsonPrimitive?)
            ?.let { FilterType.valueOf(it.content) }
        return when (filterType) {
            FilterType.GROUP -> {
                deserializeGroupFilter(json)
            }
            FilterType.HAVING_PROPERTY -> {
                deserializePropertyFilter(json)
            }
            else -> {
                throw JsonParseException("Unrecognized filter type!")
            }
        }
    }

    private fun deserializeGroupFilter(json: JsonElement): GroupFilter {
        val jsonObject = json.jsonObject
        val joinType =
            (jsonObject["joinType"] as JsonPrimitive?)?.let { JoinType.valueOf(it.content) }
                ?: throw JsonParseException("Joint type null!")
        val filters = jsonObject["filters"]?.jsonArray
        val eventFilters = mutableListOf<EventFilter>()
        filters?.let {
            for (element in it) {
                eventFilters.add(deserializeEventFilter(element))
            }
        }
        return GroupFilter(joinType, eventFilters)
    }

    private fun deserializePropertyFilter(json: JsonElement): PropertyFilter {
        val jsonObject = json.jsonObject
        val name = (jsonObject["name"] as JsonPrimitive).content
        val operationElement = jsonObject["operation"]
            ?: throw JsonParseException("operation field is null!")
        val propertyOperation = deserializeOperation(operationElement)
        return PropertyFilter(name, propertyOperation)
    }

    private fun deserializeOperation(json: JsonElement): PropertyOperation {
        val jsonObject = json.jsonObject
        val operationType = (jsonObject["type"] as JsonPrimitive?)
            ?.let { OperationType.valueOf(it.content) }
            ?: throw JsonParseException("operation type null!")
        val propertyType = (jsonObject["propertyType"] as JsonPrimitive?)
            ?.let { PropertyType.valueOf(it.content) }
            ?: throw JsonParseException("property type null!")
        return when (operationType) {
            OperationType.EQ, OperationType.GT, OperationType.LT, OperationType.GTE, OperationType.LTE, OperationType.NEQ -> {
                val value = (jsonObject["value"] as JsonPrimitive?)?.content
                    ?: throw JsonParseException("value null!")
                SingleValueOperation(value, operationType, propertyType)
            }
            OperationType.BETWEEN -> {
                val from = (jsonObject["from"] as JsonPrimitive?)?.content
                    ?: throw JsonParseException("from value null!")
                val to = (jsonObject["to"] as JsonPrimitive?)?.content
                    ?: throw JsonParseException("to value null!")
                BetweenOperation(from, to, operationType, propertyType)
            }
            else -> throw JsonParseException("Unrecognized operation type!")
        }
    }
}